design-proposals/vm-pool.md

```diff
@@ -0,0 +1,194 @@
+
+**Similarities**
+* VMConfig and VM object’s spec structure are identical
+
+## VirtualMachinePool (VMPool) API
+
+The VMPool api represents all the tunings necessary for managing a pool of stateful VMs. This API is very simple and leverages the VMConfig object for most of the heavy lifting pertaining to what the VM instances within the pool actually look like.
+
+The VMPools spec contains the following tunings and values
+
+* **VirtualMachineRef** - (Required) A structure that points to a VMConfig object
+* **Replicas** - (Required) An integer representing the desired number of VM replicas
+* **MaxUnavailable**  - (Optional) (Defaults to 25%) Integer or string pointer, that when set represents either a percentage or number of VMs in a pool that can be unavailable (ready condition false) at a time during automated update.
+* **NameGeneration** - (Optional) Specifies how objects within a pool have their names generated
+	* **VMPrefix (Defaults to VMPool.Name)** - String representing the prefix to use for when naming VMs.
+	* **AppendPostfixToSecretReferences: (default false)** - Boolean that indicates if VM’s unique postfix should be appended to references to Secrets in the VMI’s Volumes list. This is useful when needing to pre-generate unique secrets for VMs within a pool.
+	* **AppendPostfixToConfigMapReferences (default false)** - Boolean that indicates if VM’s unique postfix should be appended to ConfigMap references in the VMI’s Volumes list. This is useful when needing to pre-generate unique secrets for VMs within a pool.
+* **DisruptionStrategy** - (Optional) Specifies how the VMPool controller manages scaling in and updating VMs within a VMPool
+	* **Unmanaged** - No automation during scale-in or updates. The VM is never touched after creation. Users manually delete and update individual VMs in a pool. Persistent state preservation is up to the user removing the VMs
+	* **Opportunistic** - Opportunistic scale-in and update of VMs which are in a halted state.
+		* **ScaleInStatePreservation** - (Optional) specifies if and how to preserve state of VMs selected for scale-in.
+			* **Disabled** - (Default) all state for VMs selected for scale-in will be deleted
+			* **Offline** - PVCs for VMs selected for scale-in will be preserved and reused on scale-out (decreases provisioning time during scale out)
+			* **Online** - [NOTE we can't implement this until we have the ability to suspend VM memory state to a PVC] PVCs and memory for VMs selected for scale-in will be preserved and reused on scale-out (decreases provisioning and boot time during scale out)
+Each VM’s PVCs are preserved for future scale out
+	* **Proactive** - Proactive scale-in and update by forcing VMs to shutdown during scale-in and restart during update.
+		* **ScaleInPriority** - The priority in which VM instances are selected for proactive scale-in
+			* **Random** (Default) - Pick VMs by random for scale-in first
```

Comment:
Can you have both Proactive and ScaleInStatePreservation?  In other words, after all the stopped VMIs are removed, can the user control how the next set of VMIs are selected for removal?

Maybe `DisruptionStrategy` can be an ordered list?

Reply:
> Can you have both Proactive and ScaleInStatePreservation? 

yes, this would result in VMs which are removed during scale-in orphaning their PVCs, and then those VMs which are re-created during scale-out reclaim those orphaned PVCs.

> in other words, after all the stopped VMIs are removed, can the user control how the next set of VMIs are selected for removal?

The disruptionStrategy can be mutated at any time on the VMPool, which then impacts how future scale-in and update events occur. Is that what you're looking for?

Reply:
I had a slightly different take on policy.  I was thinking we could define one "BasePolicy", which is a policy that's always true (e.g. oldest or newest), then have a bunch of tuning policies.  "BasePolicy" would be what you have now for `Priority`.

BasePolicy - prioritize by Name or Age.  A base policy is always true so only 1 can be chosen at at time
- lowest (vm-1, vm-2, ect...)
- highest (vm-n, vm-(n-1), ect...)
- oldest
- newest
- random

TuningPolicies:
LabelPolicy - prioritize by label.  Accepts a list of labels
- Scale-in order follows the ordered list of labels. Allows admins/another controller to dynamically label VMIs to control ordering. This can be a "plug-in" for admins to inform the pool about guest information

NodePolicy - prioritize by node. Accepts a list of nodes
- Scale-in by node

Example: 
```yaml
- Priority:
    - LabelPolicy :
         - unimportant-vmis
         - not-very-important-vmis
         - important
         - super-important
     - NodePolicy:
         - node01
         - node02
     - BasePolicy: "oldest"
```
The scale-in Priority order would be:
1. VMIs labeled `unimportant-vmis`
2. VMIs labeled `not-very-important-vmis`
3. VMIs labeled `important`
4. VMIs labeled `super-important`
5. VMIs on node01
6. VMIs on node02
7. Oldest VMIs

> Maybe DisruptionStrategy can be an ordered list?
>
This is what I meant by using an ordered list.  There are a few powerful tunings we can take advance of to provide the fine-grained scale-in control.  WDYT?

Reply:
> This is what I meant by using an ordered list. There are a few powerful tunings we can take advance of to provide the fine-grained scale-in control. WDYT?

yeah, makes sense to me. Do we really need a `basePolicy` here though. For example wouldn't something like what i have below work, where there's no basePolicy but oldest is just another policy.

```
- Priority:
    - labelPolicy:
         - unimportant-vmis
         - not-very-important-vmis
         - important
         - super-important
     - nodePolicy:
         - node01
         - node02
     - OldestPolicy: {}
```

I guess the weird thing here would be if someone attempted to use the OldestPolicy as the first item in the list. That would make the rest of the items never get hit. A validation webhook could catch that though i suppose.

Reply:
Ya that would work.  The name BasePolicy is so we have a name for a group of these policies that will always match.  Maybe it needs a different name?  I just thought it was easier to communicate the point.

Reply:
> The name BasePolicy is so we have a name for a group of these policies that will always match. Maybe it needs a different name? I just thought it was easier to communicate the point.

ah that makes sense. I'll take some time to think about the naming and structure then revise the doc to take all this into account.

great feedback!

Reply:
@rthallisey I updated the doc. I called the struct `SelectionPolicy`. Within that struct, people have the option to set both a list of orderedPolicies and/or a basePolicy. 

when set, the ordered policies are matched first, in order, then if a catch all "basePolicy" is set, that will act as a catch all. If no basePolicy is set, no vms will get matched and scale-in will block until an item in the orderedPolicy matches.

Below is an example. The priority for scale-in is, by label first, then by node, then base "oldest". The priority for update in this example has no ordered list and  is simply by "oldest" 

```
apiVersion: kubevirt.io/v1
kind: VirtualMachinePool
metadata:
  name: my-vm-pool
spec:
  VirtualMachineConfigRef:
    name: my-template
  replicas: 100
  maxUnavailable: 10
  scaleInStrategy:
    proactive:
      selectionPolicy:
        orderedPolicies:
          - labelSelection
            - non-important-vms
          - nodeSelection
            - node2
            - node3
        basePolicy: "Oldest"
      statePreservation: Offline
  updateStrategy:
    proactive:
      selectionPolicy:
        basePolicy: "Oldest"
```